### PR TITLE
PR-1-TDD ruby project

### PR DESCRIPTION
**Pull request by ivanmvh**
# TDD project

## Learning objectives
- Write unit tests using Rspec.
- Understand an existing codebase in a way that allows for adding tests.
- Write Rspec unit tests with mocks.
- Understand the TDD concept.

## Description
In this project you will implement a class with some methods, but you will do it by doing test-driven development (TDD). Remember that the idea is to write tests first and then the code. This will be verified in your commits history, so it is important to use the proper sequence.

### General requirements

- Make sure that there are no linter errors.
- Make sure that you used correct Gitflow.
- Make sure that you documented your work in a professional way.

### Ruby requirements
- Follow our list of best practices for Ruby.

### Project requirements
- Create a class called `Solver`.
- Create a method called `factorial` that takes one argument, an integer `N`, and returns the [factorial](https://en.wikipedia.org/wiki/Factorial) for that number. The factorial is the multiplication of all integers from 1 to `N` and has the special case that the factorial of 0 is 1. This method only accepts 0 and positive integers, so if a negative integer is given it should raise an exception.
- Create a method called `reverse` that takes one argument, a string `word`, and returns `word` reversed (e.g. if `word` is `"hello"` it returns `"olleh"`).
- Create a method called `fizzbuzz` that takes one argument, an integer `N`, and returns a string. The returned string is constructed following these rules:
  - When `N` is divisible by 3, return `"fizz"`.
  - When `N` is divisible by 5, return `"buzz"`.
  - When `N` is divisible by 3 and 5, return `"fizzbuzz"`.
  - Any other case, return `N` as a string (e.g. say `N` is `7` then return `"7"`).
- Remember to start with tests for all of the methods described above. Your commit history should prove that you have used TDD.

@ivanmvh

### CODE REVIEW

File: spec/solver_spec.rb [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+      @solver = Solver.new
+    end
+
+    it 'Should calculate factorial of an integer number: ' do
+      expect(@solver.factorial(0)).to eq 1
+      expect(@solver.factorial(1)).to eq 1
+      expect(@solver.factorial(2)).to eq 2
+      expect(@solver.factorial(3)).to eq 6
+      expect(@solver.factorial(10)).to eq 3_628_800
+    end
```
Comment (DammyShittu) [inline spec/solver_spec.rb]: - [ ] Kindly add a test for when the number is negative (it should raise an exception).

File: spec/solver_spec.rb [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+
+    it 'Run fizzbuzz method with argument integer number and check result ' do
+      expect(@solver.fizzbuzz(93)).to eq 'fizz'
+      expect(@solver.fizzbuzz(100)).to eq 'buzz'
+      expect(@solver.fizzbuzz(30)).to eq 'fizzbuzz'
+      expect(@solver.fizzbuzz(90)).to eq 'fizzbuzz'
+      expect(@solver.fizzbuzz(14)).to eq 14
+      expect(@solver.fizzbuzz(28)).to eq 28
+    end
```
Comment (DammyShittu) [inline spec/solver_spec.rb]: - [ ] When the provided number is indivisible by 3 and 5 the tests should return the STRING version of that number. Kindly fix that.

**Review by DammyShittu — CHANGES_REQUESTED** [review]
Hi @ivanmvh and @PierreBarba,

## Highlights
✔️ Use of TDD

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!

## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Review by Deepakdanger — APPROVED** [review]
Hi @ivanmvh & @PierreBarba, 

### Status: Approved ✔️

Your project is complete!  There is nothing else to say other than... it's time to merge it  🚢 
Congratulations! 🎉 

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._
